File: libmuscle/python/libmuscle/test/test_settings_manager.py
import pytest
from ymmsl import Reference, Settings

from libmuscle.settings_manager import SettingsManager


@pytest.fixture
def settings_manager() -> SettingsManager:
    return SettingsManager()


def test_create(settings_manager):
    assert len(settings_manager.base) == 0
    assert len(settings_manager.overlay) == 0


def test_list_settings_globals(settings_manager):
    ref = Reference
    settings_manager.base[ref('test1')] = 13
    settings_manager.base[ref('test2')] = 14
    assert settings_manager.list_settings(ref('macro')) == ['test1', 'test2']


def test_list_settings_specifics(settings_manager):
    ref = Reference
    settings_manager.base[ref('macro.test1')] = 'test1'
    settings_manager.base[ref('micro.test2')] = 'test2'
    assert settings_manager.list_settings(ref('macro')) == ['test1']

    settings_manager.base[ref('micro.test1')] = 'test1'
    assert settings_manager.list_settings(ref('macro')) == ['test1']


def test_list_settings_override(settings_manager):
    ref = Reference
    settings_manager.base[ref('test1')] = 'test1'
    settings_manager.base[ref('macro.test1')] = 42
    assert settings_manager.list_settings(ref('macro')) == ['test1']

    settings_manager.base[ref('micro.test1')] = 43
    assert settings_manager.list_settings(ref('macro')) == ['test1']

    settings_manager.base[ref('test2')] = 'test2'
    assert settings_manager.list_settings(ref('macro')) == ['test1', 'test2']


def test_list_settings_overlay(settings_manager):
    ref = Reference
    settings_manager.base[ref('test1')] = 'test1'
    settings_manager.overlay[ref('test1')] = 'test1'
    assert settings_manager.list_settings(ref('macro')) == ['test1']

    settings_manager.overlay[ref('test2')] = 'test2'
    assert settings_manager.list_settings(ref('macro')) == ['test1', 'test2']


def test_list_settings_overlay_override(settings_manager):
    ref = Reference
    settings_manager.base[ref('test1')] = 'test1'
    settings_manager.base[ref('micro.test2')] = 1
    settings_manager.overlay[ref('macro.test1')] = 13
    settings_manager.overlay[ref('micro.test2')] = 2

    assert settings_manager.list_settings(ref('macro')) == ['test1']
    assert settings_manager.list_settings(ref('micro')) == ['test1', 'test2']
    assert settings_manager.list_settings(ref('meso')) == ['test1']


def test_get_setting(settings_manager):
    ref = Reference
    settings_manager.base[ref('test')] = 13
    assert settings_manager.get_setting(ref('instance'), ref('test')) == 13

    settings_manager.overlay[ref('test2')] = 14
    assert settings_manager.get_setting(ref('instance'), ref('test2')) == 14

    settings_manager.base[ref('test2')] = 'test'
    assert settings_manager.get_setting(ref('instance'), ref('test2')) == 14

    settings_manager.overlay = Settings()
    assert settings_manager.get_setting(ref('instance'), ref('test2')) == \
        'test'

    settings_manager.base[ref('test3')] = 'base_test3'
    settings_manager.base[ref('instance.test3')] = 'base_instance_test3'
    assert settings_manager.get_setting(ref('instance'), ref('test3')) == \
        'base_instance_test3'
    assert settings_manager.get_setting(ref('instance2'), ref('test3')) == \
        'base_test3'

    settings_manager.overlay[ref('test3')] = 'overlay_test3'
    settings_manager.overlay[ref('instance.test3')] = 'overlay_instance_test3'
    assert settings_manager.get_setting(ref('instance'), ref('test3')) == \
        'overlay_instance_test3'
    assert settings_manager.get_setting(ref('instance2'), ref('test3')) == \
        'overlay_test3'

    settings_manager.base[ref('instance.test4')] = 'base_test4'
    settings_manager.overlay[ref('test4')] = 'overlay_test4'
    assert settings_manager.get_setting(ref('instance'), ref('test4')) == \
        'base_test4'

    assert settings_manager.get_setting(ref('instance[10]'), ref('test4')
                                        ) == 'base_test4'

    settings_manager.base[ref('instance[10].test5')] = 'base_test5'
    settings_manager.overlay[ref('test5')] = 'overlay_test5'
    assert settings_manager.get_setting(ref('instance'), ref('test5')) == \
        'overlay_test5'
    assert settings_manager.get_setting(ref('instance[10]'), ref('test5')
                                        ) == 'base_test5'
    assert settings_manager.get_setting(ref('instance[11]'), ref('test5')
                                        ) == 'overlay_test5'


def test_get_setting_with_default(settings_manager):
    ref = Reference

    assert settings_manager.get_setting(
        ref('instance'), ref('nonexistent'), default='default_value'
    ) == 'default_value'

    assert settings_manager.get_setting(
        ref('instance'), ref('nonexistent'), default=42
    ) == 42

    assert settings_manager.get_setting(
        ref('instance'), ref('nonexistent'), default=3.14
    ) == 3.14

    assert settings_manager.get_setting(
        ref('instance'), ref('nonexistent'), default=True
    ) is True

    assert settings_manager.get_setting(
        ref('instance'), ref('nonexistent'), default=[1, 2, 3]
    ) == [1, 2, 3]

    settings_manager.base[ref('test')] = 'actual_value'
    assert settings_manager.get_setting(
        ref('instance'), ref('test'), default='default_value'
    ) == 'actual_value'

    settings_manager.overlay[ref('test2')] = 'overlay_value'
    assert settings_manager.get_setting(
        ref('instance'), ref('test2'), default='default_value'
    ) == 'overlay_value'


def test_get_setting_without_default_raises_keyerror(settings_manager):
    ref = Reference
    with pytest.raises(KeyError, match='Value for setting "nonexistent" was not set'):
        settings_manager.get_setting(ref('instance'), ref('nonexistent'))


def test_get_setting_with_default_and_type_check(settings_manager):
    ref = Reference

    result = settings_manager.get_setting(
        ref('instance'), ref('nonexistent'), typ='str', default='default'
    )
    assert result == 'default'

    settings_manager.base[ref('test')] = 42
    result = settings_manager.get_setting(
        ref('instance'), ref('test'), typ='int', default=0
    )
    assert result == 42

    with pytest.raises(TypeError):
        settings_manager.get_setting(
            ref('instance'), ref('test'), typ='str', default='default'
        )

File: src/cpp/libmuscle/tests/test_settings_manager.cpp
#include "libmuscle/settings_manager.hpp"

#include <string>
#include <vector>

#include <gtest/gtest.h>

#include <libmuscle/namespace.hpp>
#include <ymmsl/ymmsl.hpp>


using namespace std::string_literals;
using libmuscle::_MUSCLE_IMPL_NS::SettingsManager;
using ymmsl::Reference;
using ymmsl::Settings;


using VecS = std::vector<std::string>;


int main(int argc, char *argv[]) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

TEST(libmuscle_settings_manager, test_create_settings_manager) {
    SettingsManager s;

    ASSERT_EQ(s.base.size(), 0u);
    ASSERT_EQ(s.overlay.size(), 0u);
}

TEST(libmuscle_settings_manager, test_list_settings_globals) {
    SettingsManager s;

    s.base["test1"] = 13;
    s.base["test2"] = 14;
    ASSERT_EQ(s.list_settings("macro"), VecS({"test1", "test2"}));
}

TEST(libmuscle_settings_manager, test_list_settings_specifics) {
    SettingsManager s;

    s.base["macro.test1"] = "test1";
    s.base["micro.test2"] = "test2";
    ASSERT_EQ(s.list_settings("macro"), VecS({"test1"}));

    s.base["micro.test1"] = "test1";
    ASSERT_EQ(s.list_settings("macro"), VecS({"test1"}));
}

TEST(libmuscle_settings_manager, test_list_settings_override) {
    SettingsManager s;

    s.base["test1"] = "test1";
    s.base["macro.test1"] = 42;
    ASSERT_EQ(s.list_settings("macro"), VecS({"test1"}));

    s.base["micro.test1"] = 43;
    ASSERT_EQ(s.list_settings("macro"), VecS({"test1"}));

    s.base["test2"] = "test2";
    ASSERT_EQ(s.list_settings("macro"), VecS({"test1", "test2"}));
}

TEST(libmuscle_settings_manager, test_list_settings_overlay) {
    SettingsManager s;

    s.base["test1"] = "test1";
    s.overlay["test1"] = "test1";
    ASSERT_EQ(s.list_settings("macro"), VecS({"test1"}));

    s.overlay["test2"] = "test2";
    ASSERT_EQ(s.list_settings("macro"), VecS({"test1", "test2"}));
}

TEST(libmuscle_settings_manager, test_list_settings_overlay_override) {
    SettingsManager s;

    s.base["test1"] = "test1";
    s.base["micro.test2"] = 1;
    s.overlay["macro.test1"] = 13;
    s.overlay["micro.test2"] = 2;

    ASSERT_EQ(s.list_settings("macro"), VecS({"test1"}));
    ASSERT_EQ(s.list_settings("micro"), VecS({"test1", "test2"}));
    ASSERT_EQ(s.list_settings("meso"), VecS({"test1"}));
}

TEST(libmuscle_settings_manager, test_get_setting) {
    SettingsManager s;

    s.base["test"] = 13;
    ASSERT_EQ(s.get_setting("instance"s, "test"s), 13);

    s.overlay["test2"] = 14;
    ASSERT_EQ(s.get_setting("instance"s, "test2"s), 14);

    s.base["test2"] = "test";
    ASSERT_EQ(s.get_setting("instance"s, "test2"s), 14);

    s.overlay = Settings();
    ASSERT_EQ(s.get_setting("instance"s, "test2"s), "test");


    s.base["test3"] = "base_test3";
    s.base["instance.test3"] = "base_instance_test3";
    ASSERT_EQ(s.get_setting("instance"s, "test3"s), "base_instance_test3");
    ASSERT_EQ(s.get_setting("instance2"s, "test3"s), "base_test3");

    s.overlay["test3"] = "overlay_test3";
    s.overlay["instance.test3"] = "overlay_instance_test3";
    ASSERT_EQ(s.get_setting("instance"s, "test3"s), "overlay_instance_test3");
    ASSERT_EQ(s.get_setting("instance2"s, "test3"s), "overlay_test3");

    s.base["instance.test4"] = "base_test4";
    s.overlay["test4"] = "overlay_test4";
    ASSERT_EQ(s.get_setting("instance"s, "test4"s), "base_test4");
    ASSERT_EQ(s.get_setting("instance[10]"s, "test4"s), "base_test4");

    s.base["instance[10].test5"] = "base_test5";
    s.overlay["test5"] = "overlay_test5";
    ASSERT_EQ(s.get_setting("instance"s, "test5"s), "overlay_test5");
    ASSERT_EQ(s.get_setting("instance[10]"s, "test5"s), "base_test5");
    ASSERT_EQ(s.get_setting("instance[11]"s, "test5"s), "overlay_test5");
}

